src/java/voldemort/client/protocol/admin/AdminClient.java

```diff
@@ -299,6 +301,28 @@ public void close() {
     }
 
     /**
+     * If Cluster is modified ( nodes are added/removed), using the old
+     * AdminClient will cause inconsistent operations. This method returns true
+     * under such cases, so that caller can discard the current AdminClient and
+     * create new AdminClient.
+     * 
+     * @return true if cluster is modified since the AdminClient is create,
+     *         false otherwise
+     */
+    public boolean isClusterModified() {
```

Comment:
For AdminCleint to be self sufficient, it needs to do the following
1) Refresh cluster.xml
2) Flush all the connections and reset all its state to new one.
3) Some of the ongoing operation, may need to be re-attempted or retried.

ZenStoreClient does that to cope with cluster replacement and cluster failure. But AdminClient is designed with the assumption that a cluster must be healthy or the command will fail.

Doing the above 3 is very tough if not impossible. Exposing the complexity to the caller is compromise, but will work in the short term. AdminClient is designed for operator and making it run on a self recovering API environment is a very big task.

Thoughts/Opinions ?